openhands/server/routes/files.py

```diff
@@ -77,10 +77,16 @@ async def list_files(request: Request, path: str | None = None):
 
     file_list = [f for f in file_list if f not in FILES_TO_IGNORE]
 
-    async def filter_for_gitignore(file_list, base_path):
-        gitignore_path = os.path.join(base_path, '.gitignore')
+    async def has_gitignore() -> bool:
+        file_list = await call_sync_from_async(runtime.list_files, '')
+        result = next((True for f in file_list if f == '.gitignore'), False)
+        return result
+
+    async def filter_for_gitignore(file_list):
+        if not await has_gitignore():
+            return file_list
         try:
-            read_action = FileReadAction(gitignore_path)
+            read_action = FileReadAction('.gitignore')
             observation = await call_sync_from_async(runtime.run_action, read_action)
```

Comment:
Just a thought, if there is no addition to the `stream` on this execution path (and it shouldn't be), could we just get/read the ErrorObs and treat it here directly?

Reply:
With this approach, the logs would still contain:
```
|**ErrorObservation**
    |File not found: /workspace/.gitignore. Your current working directory is /workspace
```
The only ways around this that I can see are:

1. Live with this oddity.
2. Do not filter files using `.gitignore`
3. Check that the `.gitignore` exists. (The approach I took)
4. Update the `FileReadAction` with an optional flag to permit missing files and not yield an error observation.
5. Update the runtime to allow access to the FileStore and use this directly.

Reply:
Correct me if wrong, but isn't there also
6. Do not log stuff in the FileReadAction, but in its client code?

Or that sounds really bad? I don't think they are many... but I could be wrong.

Reply:
Actually, [this one](https://github.com/All-Hands-AI/OpenHands/blob/2f11634ccaeb3aa38893a994d53ff280bbf485d0/openhands/runtime/action_execution_server.py#L254) isn't doing it, it's the `run_action`.

Reply:
How about we use `runtime.read` directly instead of `run_action`

Reply:
Looks like it also does `run_action`. We could write one that doesn't, unless I'm missing something?

OK, but it's late and sorry, I no longer understand this problem in the first place: these are all in the sandbox, both the results of `list_files` and `.gitignore`, so why aren't we filtering during `runtime.list_files`?